flybywiresim/aircraft
# refactor(mcdu): custom scratchpad

## Summary of Changes

This scratchpad features all previous features plus a few new as well as a great foundation for more advanced scenarios.

Primarily you will observe a new behavior of the scratchpad (which is true to life) which wasn't featured before.

#### What is new?

This version is build from scratch and treated as a dedicated "system".

#### What has changed?

The MCDUs subsystems now need to take care of telling the scratchpad whether to display something or not, especially when entering data via the line select keys the corresponding subsystem now is in charge to send feedback to the scratchpad if the input is e.g. invalid.

Due to the absence of a uniform input validation process integrating the new callback (contains a function to be called in case the user made a bad input) required a manual call to be added to various validations.





## References

See pilot feedback and fms channel for preview.




## Additional context



Discord username (if different from GitHub): DerL30N#3751

## Testing instructions

- Ensure previous functionalities are still present (nothing has changed in terms of data input and manipulation)
- Ensure whenever an input into the MCDU is made via the LSK and you got a message for bad input, that the data you tried to input is present on the scratchpad after you've cleared the "error" message.
_(Note that this is not true for `NOT YET IMPLEMENTED` message, since you'd have to clear the content one way or the other)_

At best try to enter data into every page in order to find issues or missing functions.

_Don't worry if a type II message "flickers" in/out after you've input data via LSK and revived a message for bad input, this is intended behavior, and true to life._


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
